ksql-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -331,6 +334,94 @@ public Void visitSubscriptExpression(
       return null;
     }
 
+    @Override
+    public Void visitCreateArrayExpression(
+        final CreateArrayExpression exp,
+        final ExpressionTypeContext context
+    ) {
+      if (exp.getValues().isEmpty()) {
+        throw new KsqlException(
+            "Array constructor cannot be empty. Please supply at least one element "
+                + "(see https://github.com/confluentinc/ksql/issues/4239).");
+      }
+
+      final List<SqlType> sqlTypes = exp
+          .getValues()
+          .stream()
+          .map(val -> {
+            process(val, context);
+            return context.getSqlType();
+          })
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+
+      if (sqlTypes.size() == 0) {
+        throw new KsqlException("Cannot construct an array with all NULL elements "
+            + "(see https://github.com/confluentinc/ksql/issues/4239). As a workaround, you may "
+            + "cast the NULL value to a desired type.");
+      }
+
+      if (new HashSet<>(sqlTypes).size() != 1) {
+        throw new KsqlException(
+            String.format(
+                "Cannot construct an array with mismatching types (%s) from expression %s.",
+                sqlTypes,
+                exp));
+      }
+
+      context.setSqlType(SqlArray.of(sqlTypes.get(0)));
+      return null;
+    }
+
+    @Override
+    public Void visitCreateMapExpression(
+        final CreateMapExpression exp,
+        final ExpressionTypeContext context
+    ) {
+      if (exp.getMap().isEmpty()) {
+        throw new KsqlException("Map constructor cannot be empty. Please supply at least one key "
+            + "value pair (see https://github.com/confluentinc/ksql/issues/4239).");
+      }
+
+      final List<SqlType> gkeyTypes = exp.getMap()
+          .keySet()
+          .stream()
+          .map(key -> {
+            process(key, context);
+            return context.getSqlType();
+          })
+          .collect(Collectors.toList());
+
+      if (gkeyTypes.stream().anyMatch(type -> !SqlTypes.STRING.equals(type))) {
+        throw new KsqlException("Only STRING keys are supported in maps but got: " + gkeyTypes);
+      }
+
+      final List<SqlType> sqlTypes = exp.getMap()
```

Comment:
super nit: can we call this `valueTypes` and the other `keyTypes` to make it clear?